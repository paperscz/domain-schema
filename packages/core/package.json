{
  "name": "@domain-schema/core",
  "version": "0.0.34",
  "description": "A JavaScript Schema for Domain Driven Design",
  "repository": "https://github.com/sysgears/domain-schema.git",
  "author": "Victor Vlasenko <victor.vlasenko@sysgears.com>",
  "main": "lib/index.js",
  "license": "MIT",
  "scripts": {
    "clean": "rm -rf ./lib",
    "build": "tsc -p tsconfig.build.json",
    "watch": "tsc -p tsconfig.build.json -w",
    "lint": "tslint -p tsconfig.json --fix -c ../../tslint.json",
    "tests": "jest",
    "test": "yarn tests && yarn lint",
    "prepublishOnly": "yarn clean && yarn build"
  },
  "files": [
    "lib/*.js",
    "lib/*.d.ts",
    "lib/*.map"
  ],
  "dependencies": {
    "debug": "^3.1.0"
  },
  "jest": {
    "transform": {
      ".(ts|tsx)": "ts-jest"
    },
    "testRegex": "(/__tests__/.*\\.(test|spec))\\.(ts|tsx|js)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "json"
    ],
    "testEnvironment": "node"
  }
}
