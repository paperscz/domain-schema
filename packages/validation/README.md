# Domain Schema Validation

[![npm version](https://badge.fury.io/js/domain-react-forms.svg)](https://badge.fury.io/js/domain-react-forms) [![Twitter Follow](https://img.shields.io/twitter/follow/sysgears.svg?style=social)](https://twitter.com/sysgears)

## Installation

```bash
yarn add @domain-schema/validation
```

or

```bash
npm install @domain-schema/validation
```

## Validation

Domain Validation intended to check values whether they comply with structure of
 domain schema fields.

### Usage
Firstly import ```DomanValidator``` from domain-schema validation package
```js
    import DomainValidator from '@domain-schema/validation'
```
Then call ```DomanValidator``` static method ```validate``` 
```js
    DomainValidator.validate(schema, values)
```
With such params:
   * ```schema``` - ```DomainSchema``` instance
   * ```values``` - Formatted object from form or another source with values which 
    ```DomainValidator``` will check on conformity of ```DomainSchema```
      * key - Field name correspond to schema field
      * value - Entity which must pass validation
         ```js 
          { name : "Jack" }
        ```
You can register domain-schema field validation using following approaches:
   *  Type defining
       ```js 
        isAdmin: {
            type: Boolean
        }
       ```
   * Build-in validators
   * Custom validators

For more accurate validation you could use several approaches simultaneously

### Built-in validators

* ```optional``` - By default, all keys are required. Set ```optional: true``` to change that.

  ```js
    name = {
      ...
      optional: true,
      ...
    }
  ```
  * ```matches``` - Checks if the value matches some specific field

  ```js
    password = {
      ...
    };
    passwordConfirmation = {
      ...
      matches: 'password',
      ...
    }
  ```

  * ```max``` - Checks if the value or value length does not exceed a specified number
   (works with both numbers and strings)

  ```js
    // Checks length if type is String...
    name = {
      ...
      type: String,
      max: 6,
      ...
    }
    // ...and value when type is Number
    age = {
      ...
      type: Number,
      max: 16,
      ...
    }
  ```

  * ```min``` - Checks if the value or value length not less then a specified number
  (works with both numbers and strings)

  ```js
    // Checks length if type is String...
    name = {
      ...
      type: String,
      min: 6,
      ...
    }
    // ...and value when type is Number
    age = {
      ...
      type: Number,
      min: 16,
      ...
    }
  ```

  * ```email``` - Checks if the value corresponds to an email

  ```js
    name = {
      ...
      email: true,
      ...
    }
  ```

  * ```alphaNumeric``` - Checks if the value consists of alphanumeric characters

  ```js
    text = {
      ...
      alphaNumeric: true,
      ...
    }
  ```

  * ```phoneNumber``` - Checks if the value corresponds to a phone number

  ```js
    phone = {
      ...
      phoneNumber: true,
      ...
    }
  ```

  * ```equals``` - Checks if the value equals to a specific value

  ```js
    role = {
      ...
      equals: 'admin',
      ...
    }
  ```

### Nested Schema Validation
Be careful when establish relation via schema field on your domain schema
```js
class Product extends Schema {
  __ = { name: 'Product', tablePrefix: '' };
  id = DomainSchema.Int;
  name = {
    type: String,
    searchText: true
  };
  category = {
    type: Category
   };
}
```

 ```js
class Category extends Schema {
  __ = { name: 'Category', tablePrefix: '' };
  id = DomainSchema.Int;
  name = {
    type: String,
    searchText: true,
  };
  products = {
    type: [Product]
  };
}
```
In this case ```domainValidator``` also will check all fields of nested Category
 schema and then Product again.

So object must have the following structure in order to pass validation
```js
{
  id : 1,
  name : 'Iphone X'
  category : {
    id : 1,
    name : 'Phones',
    products : ...
  }
}
```
To prevent check of nested schema use ```blackbox``` attribute
```js
  products = {
    type: [Product],
    blackbox: true
  };
```
Now the following object will pass validation
```js
{
  id : 1,
  name : 'Iphone X'
  category : {
    id : 1,
    name : 'Phones'
  }
}
```


### Customizing Validation Messages

Validation error messages can be overridden by defining messages in one place as a single
object and passing it to the ```setValidationMessages``` function.
In that object the ``keys`` are validator names and values can be either strings or functions, as follows:

* Values are strings:

```js
    import DomainValidation from '@domain-schema/validation';

    DomainValidation.setValidationMessages({
      required: 'This field is required',
      phoneNumber: 'Error! Not a phone number!'
    });
```

* Values are callback functions:

```js
    import DomainValidation from '@domain-schema/validation';

    DomainValidation.setValidationMessages({
      required: ({fieldName}) => {
        return `Field '${fieldName}' is required`
      },
      phoneNumber: ({values, fieldName}) => {
        return `Error! '${values[fieldName]}' is not a phone number`
      }
    });
```

Callback functions, in turn, get object with the following properties:

* ``values``    - the validation object
* ``fieldName`` - the current field name
* ``schema``    - the domain schema definition object

We can also define a custom validation error message for a **specific field** right in the schema:

```js
    ...
    name = {
        type: String,
        ...
        required: {
          value: true,
          msg: 'Required Name'
        }
    }
    ...
```

### Custom validators

A user is able to add any number of custom validators whenever it need by passing functions to the ```validators```
property of a schema field:

```js
    password = {
        type: String,
        ...
        validators: [(value, values) => {
          return value.length > 3 ? undefined : 'Must Be more than 3 characters';
        }]
    };
```

  Validation callback function gets the following params:
* ``value`` - the value for being validated
* ``values`` - the form object

NOTE: validation function must return a string with an error message or ```undefined```.

## License

Copyright Â© 2018 [SysGears INC]. This source code is licensed under the [MIT] license.

[MIT]: LICENSE
[SysGears INC]: http://sysgears.com
[View docs]: https://github.com/sysgears/domain-schema/blob/master/packages/formik/README.md
