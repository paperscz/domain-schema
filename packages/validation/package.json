{
  "name": "@domain-schema/validation",
  "version": "0.0.36",
  "description": "Validation for Domain Schema",
  "repository": "https://github.com/sysgears/domain-schema.git",
  "author": "Victor Vlasenko <victor.vlasenko@sysgears.com>",
  "main": "src/index.ts",
  "license": "MIT",
  "scripts": {
    "clean": "rm -rf ./lib",
    "build": "tsc -p tsconfig.build.json",
    "watch": "tsc -p tsconfig.build.json -w",
    "prepublishOnly": "yarn clean && yarn build"
  },
  "files": [
    "lib/*.js",
    "lib/*.d.ts",
    "lib/*.map"
  ],
  "devDependencies": {
    "@domain-schema/core": ">= 0.0.30",
    "@types/node": "^7.0.18"
  },
  "peerDependencies": {
    "@domain-schema/core": ">= 0.0.30"
  },
  "dependencies": {
    "debug": "^3.1.0",
    "humps": "^2.0.1"
  },
  "publishConfig": {
    "main": "lib/index.js",
    "types": "lib/index.d.js"
  }
}
