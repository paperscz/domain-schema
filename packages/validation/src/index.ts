import DomainSchema, { Schema } from '@domain-schema/core';
import { Condition, SchemaContext, Value } from './types';
import validators, { supportedValidators } from './validators';

export default class DomainValidator {
  /**
   *
   * @type {([BooleanConstructor , string] | [NumberConstructor , string] | [StringConstructor , string] |
   * [DateConstructor , string])[]}
   */
  public static readonly TYPE_VALIDATORS: ReadonlyArray<[object, string]> = [
    [Boolean, 'bool'],
    [Number, 'numeric'],
    [String, 'string'],
    [Date, 'date']
  ];

  constructor() {}

  /**
   *
   * @param {} initialSchema
   * @param values
   * @returns {any}
   */
  public validate(initialSchema: Schema, values: any = {}): any {
    const domainSchema = new DomainSchema(initialSchema);
    return this.validateSchema(domainSchema, values, []);
  }

  /**
   *
   * @param {} schema
   * @param fieldName
   * @param schemaField
   * @param value
   * @param values
   * @returns {any}
   */
  public validateField(schema: DomainSchema, fieldName: any, schemaField: any, value: any, values: any): any {
    const error = { [fieldName]: [] };
    if (!schemaField.optional) {
      const result = this.checkWithValidator(
        'required',
        value,
        { values, fieldName, schema },
        schemaField.required || true
      );
      if (result) {
        error[fieldName].push(result);
      }
    }
    const hasTypeOf = targetType => schemaField.type === targetType || schemaField.type.prototype instanceof targetType;

    // validation for schema type
    for (const validator of DomainValidator.TYPE_VALIDATORS) {
      const [type, rule] = validator;
      if (hasTypeOf(type) && supportedValidators[rule]) {
        const result = this.checkWithValidator(rule, value, { values, fieldName, schema }, '');
        if (result) {
          error[fieldName].push(result);
        }
      }
    }

    // validation for additional validators
    Object.keys(schemaField).forEach((key: string) => {
      if (supportedValidators[key] && key !== 'type') {
        const result = this.checkWithValidator(key, value, { values, fieldName, schema }, schemaField[key]);
        if (result) {
          error[fieldName].push(result);
        }
      } else if (key === 'validators') {
        // handling custom validators
        schemaField[key].forEach(validator => {
          const receivedError = validator(value, values);
          if (receivedError) {
            error[fieldName].push(receivedError);
          }
        });
      }
    });

    if (error[fieldName] && Array.isArray(error[fieldName])) {
      if (error[fieldName].length === 0) {
        return {};
      } else if (error[fieldName].length === 1) {
        return { [fieldName]: error[fieldName][0] };
      }
    }

    return error;
  }

  /**
   *
   * @param {} schema
   * @param values
   * @param {string[]} seen
   * @returns {any}
   */
  private validateSchema(schema: DomainSchema, values: any, seen: string[]): any {
    const errors = {};
    seen.push(schema.__.name);
    for (const key of schema.keys()) {
      const schemaField = schema.values[key];
      const value = values ? (values[key] ? values[key] : null) : null;
      let nestedSchema = null;
      if (schemaField.type.isSchema) {
        nestedSchema = schemaField.type;
      } else if (schemaField.type.constructor === Array) {
        nestedSchema = schemaField.type[0];
      }
      if (nestedSchema) {
        if (seen.indexOf(nestedSchema.__.name) >= 0 || schemaField.blackbox) {
          continue;
        }
        const nestedSchemaErrors = this.validateSchema(nestedSchema, value, seen);
        if (Object.keys(nestedSchemaErrors).length === 0 && nestedSchemaErrors.constructor === Object) {
          continue;
        }
        Object.assign(errors, { [key]: nestedSchemaErrors });
      } else {
        Object.assign(errors, this.validateField(schema, key, schemaField, value, values));
      }
    }

    return errors;
  }

  /**
   *
   * @param {string} validatorName
   * @param {Value} value
   * @param {SchemaContext} context
   * @param {Condition} condition
   * @returns {any}
   */
  private checkWithValidator(validatorName: string, value: Value, context: SchemaContext, condition: Condition) {
    return typeof condition !== 'object'
      ? validators[validatorName](value)(context, condition)
      : validators[validatorName](value, condition.msg)(context, condition.value);
  }

  /**
   * @param messages
   */
  public static setValidationMessages(messages: any) {
    validators.setValidationMsg(messages);
  }
}
