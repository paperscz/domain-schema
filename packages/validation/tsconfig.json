{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "lib",
    "downlevelIteration": true
  },
  "include": ["src/**/*"]
}
