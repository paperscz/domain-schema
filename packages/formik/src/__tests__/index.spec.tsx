import { Schema } from '@domain-schema/core';
import { FieldValidators } from '@domain-schema/validation';
import * as React from 'react';
import renderer from 'react-test-renderer';

import  UserTestSchema1 from './schema/testSchema1';
import { User as UserTestSchema2 } from './schema/testSchema2';
import  UserTestSchema3 from './schema/testSchema3';
import { DomainSchemaFormik, FSF } from '../';
import { Button, Form, RenderCheckBox, RenderField, RenderSelect, RenderSelectQuery } from './components';

DomainSchemaFormik.setFormComponents({
  input: RenderField,
  checkbox: RenderCheckBox,
  select: RenderSelect,
  button: Button,
  form: Form
});

describe('DomainFormik', () => {
  it('should generate simple form', () => {
    const form = new DomainSchemaFormik(UserTestSchema1);
    const FormComponent = form.generateForm();
    const component = renderer.create(<FormComponent onSubmit={() => null} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should generate complex form', () => {
    const userForm = new DomainSchemaFormik(UserTestSchema2);
    userForm.setFormComponents({
      select: RenderSelectQuery
    });
    const FormComponent = userForm.generateForm({
      label: 'Submit',
      disableOnInvalid: false
    });
    const component = renderer.create(<FormComponent onSubmit={() => null} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should generate form with custom field', () => {
    const MyComponent = ({ input: { name, value, className, ...attrs }, label }) => {
      return (
        <div>
          <input {...attrs} type="text" name={name} placeholder={label} value={value} className={className} />
        </div>
      );
    };
    const schema = class User extends Schema {
      public __ = { name: 'User' };
      public username: FSF = {
        type: String,
        input: {
          label: 'Username'
        },
        defaultValue: 'User'
      };
      public userfield: FSF = {
        type: String,
        fieldType: 'custom',
        component: MyComponent,
        input: {
          label: 'User Field',
          className: 'my-awesome-field'
        },
        defaultValue: 'my field'
      };
    };

    const userForm = new DomainSchemaFormik(schema);
    const FormComponent = userForm.generateForm({
      label: 'Submit'
    });
    const component = renderer.create(<FormComponent onSubmit={() => null} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should allow to set UI toolkit for particular instance', () => {
    const MyInputComponent = ({ input: { name, value, className, ...attrs }, label }) => {
      return (
        <div>
          <input
            {...attrs}
            type="text"
            name={'custom_' + name}
            placeholder={label}
            value={value}
            className={className}
          />
        </div>
      );
    };
    const form = new DomainSchemaFormik(UserTestSchema3);
    form.setFormComponents({
      input: MyInputComponent
    });
    const FormComponent = form.generateForm();
    const component = renderer.create(<FormComponent onSubmit={() => null} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
