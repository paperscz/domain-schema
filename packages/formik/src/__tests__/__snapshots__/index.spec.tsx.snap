// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomainFormik should allow to set UI toolkit for particular instance 1`] = `
<form
  className=""
  name="User"
>
  <div>
    <input
      className={undefined}
      name="custom_username"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value="John"
    />
  </div>
  <div>
    <input
      className={undefined}
      name="custom_email"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
  </div>
  <div>
    <input
      className={undefined}
      name="custom_pass"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Password"
      type="text"
      value=""
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <button
      className="btn btn-secondary disabled"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Save
    </button>
  </div>
</form>
`;

exports[`DomainFormik should generate complex form 1`] = `
<form
  className=""
  name="User"
>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="name"
    >
      Username
    </label>
    <input
      className="form-control"
      id="name"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value="User"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="email"
    >
      Email
    </label>
    <input
      className="form-control"
      id="email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="User email"
      type="email"
      value=""
    />
  </div>
  <div>
    <div>
      <select
        className="form-control"
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "100%",
          }
        }
        type="select"
        value="0"
      >
        <option
          value="1"
        >
          Group_1
        </option>
        <option
          value="2"
        >
          Group_2
        </option>
        <option
          value="3"
        >
          Group_3
        </option>
        <option
          value="4"
        >
          Group_4
        </option>
        <option
          value="5"
        >
          Group_5
        </option>
      </select>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="password"
    >
      Password
    </label>
    <input
      className="form-control"
      id="password"
      name="password"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="passwordConfirmation"
    >
      Password Confirmation
    </label>
    <input
      className="form-control"
      id="passwordConfirmation"
      name="passwordConfirmation"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Password Confirmation"
      type="password"
      value=""
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <button
      className="btn btn-secondary"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`DomainFormik should generate form with custom field 1`] = `
<form
  className=""
  name="User"
>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="username"
    >
      Username
    </label>
    <input
      className="form-control"
      id="username"
      name="username"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value="User"
    />
  </div>
  <div>
    <input
      className="my-awesome-field"
      name="userfield"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="User Field"
      type="text"
      value="my field"
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <button
      className="btn btn-secondary disabled"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`DomainFormik should generate simple form 1`] = `
<form
  className=""
  name="User"
>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="username"
    >
      Username
    </label>
    <input
      className="form-control"
      id="username"
      name="username"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="name"
      type="text"
      value="John"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="email"
    >
      Email
    </label>
    <input
      className="form-control"
      id="email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="pass"
    >
      Email
    </label>
    <input
      className="form-control"
      id="pass"
      name="pass"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Email"
      type="password"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <div
      className="form-check"
    >
      <input
        checked={true}
        className="form-check-input"
        id="isActive"
        name="isActive"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="form-check-label "
        htmlFor="isActive"
      >
        Is Active
      </label>
    </div>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <button
      className="btn btn-secondary disabled"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Save
    </button>
  </div>
</form>
`;
